Fix/crash missing mybe

bparli opened the pull request:
This PR is meant to address #2460 

If there is a more appropriate way to fix this condition just let me know and I'm happy to work on it.

Commit:
don't crash on missing backend

Commit:
spacing

pondix commented:
Automated message: PR pending admin approval for build testing

renecannao commented:
Hi @bparli .
Do you have a core dump of a crash?
A MySQL_Session without `client_myds` and without `mybe` at the same time should never exist, thus the `assert()`

bparli commented:
Hi @renecannao - we're getting bit by the `assert(mybe)` line.  I probably misunderstood `client_myds` then

I'll try getting a core dump today and attach to the issue

Commit:
fix AND logic

bskllzh commented:
@bparli , i think your fix code has a problem, for example, when mybe or mybe->server_myds object is null,  proxysql will continue to execute go down. Howerver，this will lead to use the null pointer that will cause a new coredump.

the fowllow code from your fix:
	//	assert(mybe);
	//	assert(mybe->server_myds);
	//	goto handler_again;
                if (mybe && mybe->server_myds) {
                  goto handler_again;
                }

bparli commented:
Hi @renecannao - our Security folks aren't keen on sharing the entire core dump but I've extracted some relevant bits and added to the #2460 issue.  It does appear both `client_myds` and `mybe` (and even `mybes`) are NULL

bparli committed:
return -1 instead of assert

bparli commented:
Updated the PR to the patch we're running.  We've had this for a week in one of our environments and not seen any crashes.  I found I also had to set `mysql-reset_connection_algorithm=1` in order to avoid subsequent segfaults

mirostauder commented:
Can one of the admins verify this patch?

renecannao commented:
Closing this PR as obsolete. This crash fix from 2020 (4+ years old) addresses issue #2460, which was already resolved and closed in January 2024 through other fixes. The crash bug described in #2460 has been fixed through subsequent development, making this specific approach no longer needed. Thank you @bparli for the contribution!
